This is an introduction and tutorial to the SQL Clause WITH, also known as a Common Table Expression or CTE. If you have a good handle of writing SQL queries with complex joins and are at least familiar with writing subqueries, you should find this helpful!

To me, there is nothing more annoying (and by this I mean mildly infuriating and embarrassing) than when I return to my own code and can’t remember why I wrote what I did and what I was trying to accomplish. SQL code can be very straightforward as a language, but introducing subqueries and nested subqueries quickly obfuscates the intention. 

The WITH clause allows us to introduce a query or set of queries that even presents deeply nested subqueries to be easily followed and applied to the parent query. The WITH clause introduces what most people refer to as Common Table Expressions, or CTEs. 

I will be using the California water quality standards and goals, last update in January, 2023. The original .pdf and the cleaned .csv file are available on GitHub: [https://github.com/FreshOats/Water\_Data\_Tutorials](https://github.com/FreshOats/Water_Data_Tutorials). 

The file I will be using for the database is the state\_regulations.csv  
The file for MS SQL Server users to create and populate a database is the state\_regulations\_database.sql file. 

Once you have established the database, let’s run a quick query:

In [1]:
SELECT * 
FROM dbo.state_regulations

(91 rows affected)

Total execution time: 00:00:00.012

contaminant state_max_level state_detection_limit state_health_goal state_health_date federal_max_level federal_max_level_goal units Aluminum 1 0.05 0.6 2001 NULL NULL mg/L Antimony 0.006 0.006 0.001 2016 0.006 0.006 mg/L Arsenic 0.01 0.002 4E-06 2004 0.01 0 mg/L Asbestos 7 0.2 7 2003 7 7 MFL Barium 1 0.1 2 2003 2 2 mg/L Beryllium 0.004 0.001 0.001 2003 0.004 0.004 mg/L Cadmium 0.005 0.001 4E-05 2006 0.005 0.005 mg/L Chromium, Total 0.05 0.01 NULL 1999 0.1 0.1 mg/L Cyanide 0.15 0.1 0.15 1997 0.2 0.2 mg/L Fluoride 2 0.1 1 1997 4 4 mg/L Mercury 0.002 0.001 0.0012 2005 0.002 0.002 mg/L Nickel 0.1 0.01 0.012 2001 NULL NULL mg/L Nitrate 10 0.4 45 2018 10 10 10 as N mg/L Nitrite 1 0.4 1 2018 1 1 1 as N mg/L Nitrate + Nitrite 10 NULL 10 2018 NULL NULL 10 as N mg/L Perchlorate 0.006 0.002 0.001 2015 NULL NULL mg/L Selenium 0.05 0.005 0.03 2010 0.05 0.05 mg/L Thallium 0.002 0.001 0.0001 2004 0.002 0.0005 mg/L Copper 1.3 0.05 0.3 2008 1.3 1.3 mg/L Lead 0.015 0.005 0.0002 2009 0.015 0 mg/L Gross Alpha Particle 15 3 NULL NULL 15 0 pCi/L Gross Beta Particle 4 4 NULL NULL 4 0 pCi/L Radium-226 + Radium-228 5 NULL NULL NULL 5 0 pCi/L Strontium-90 8 2 0.35 2006 NULL NULL pCi/L Tritium 20000 1000 400 2006 NULL NULL pCi/L Uranium 20 1 0.43 2001 30 0 pCi/L (ug/L for Federal_MCL) Benzene 0.001 0.0005 0.00015 2001 0.005 0 mg/L Carbon tetrachloride 0.0005 0.0005 0.0001 2000 0.005 0 mg/L 1,2-Dichlorobenzene 0.6 0.0005 0.6 2009 0.6 0.6 mg/L 1,4-Dichlorobenzene(p-DCB) 0.005 0.0005 0.006 1997 0.075 0.075 mg/L 1,1-Dichloroethane (1,1-DCA) 0.005 0.0005 0.003 2003 NULL NULL mg/L 1,2-Dichloroethane (1,2-DCA) 0.0005 0.0005 0.0004 2005 0.005 0 mg/L 1,1-Dichloroethylene (1,1-DCE) 0.006 0.0005 0.01 1999 0.007 0.007 mg/L cis-1,2-Dichloroethylene 0.006 0.0005 0.013 2018 0.07 0.07 mg/L trans-1,2-Dichloroethylene 0.01 0.0005 0.05 2018 0.1 0.1 mg/L Dichloromethane (Methylene chloride) 0.005 0.0005 0.004 2000 0.005 0 mg/L 1,2-Dichloropropane 0.005 0.0005 0.0005 1999 0.005 0 mg/L 1,3-Dichloropropene 0.0005 0.0005 0.0002 2006 NULL NULL mg/L Ethylbenzene 0.3 0.0005 0.3 1997 0.7 0.7 mg/L Methyl tertiary butyl ether (MTBE) 0.013 0.003 0.013 1999 NULL NULL mg/L Monochlorobenzene 0.07 0.0005 0.07 2014 0.1 0.1 mg/L Styrene 0.1 0.0005 0.0005 2010 0.1 0.1 mg/L 1,1,2,2-Tetrachloroethane 0.001 0.0005 0.0001 2003 0.1 0.1 mg/L Tetrachloroethylene (PCE) 0.005 0.0005 6E-05 2001 0.005 0 mg/L Toluene 0.15 0.0005 0.15 1999 1 1 mg/L 1,2,4-Trichlorobenzene 0.005 0.0005 0.005 1999 0.07 0.07 mg/L 1,1,1-Trichloroethane (1,1,1-TCA) 0.2 0.0005 1 2006 0.2 0.2 mg/L 1,1,2-Trichloroethane (1,1,2-TCA) 0.005 0.0005 0.0003 2006 0.005 0.003 mg/L Trichloroethylene (TCE) 0.005 0.0005 0.0017 2009 0.005 0 mg/L Trichlorofluoromethane (Freon 11) 0.15 0.005 1.3 2014 NULL NULL mg/L 1,1,2-Trichloro-1,2,2-Trifluoroethane (Freon 113) 1.2 0.01 4 2011 NULL NULL mg/L Vinyl chloride 0.0005 0.0005 5E-05 2000 0.002 0 mg/L Xylenes 1.75 0.0005 1.8 1997 10 10 mg/L Alachlor 0.002 0.001 0.004 1997 0.002 0 mg/L Atrazine 0.001 0.0005 0.00015 1999 0.003 0.003 mg/L Bentazon 0.018 0.002 0.2 2009 NULL NULL mg/L Benzo(a)pyrene 0.0002 0.0001 7E-06 2010 0.0002 0 mg/L Carbofuran 0.018 0.005 0.0007 2016 0.04 0.04 mg/L Chlordane 0.0001 0.0001 3E-05 2006 0.002 0 mg/L Dalapon 0.2 0.01 0.79 2009 0.2 0.2 mg/L 1,2-Dibromo-3-chloropropane (DBCP) 0.0002 1E-05 3E-06 2020 0.0002 0 mg/L 2,4-Dichlorophenoxyacetic acid (2,4-D) 0.07 0.01 0.02 2009 0.07 0.07 mg/L Di(2-ethylhexyl)adipate 0.4 0.005 0.2 2003 0.4 0.4 mg/L Di(2-ethylhexyl)phthalate (DEHP) 0.004 0.003 0.012 1997 0.006 0 mg/L Dinoseb 0.007 0.002 0.014 2010 0.007 0.007 mg/L Diquat 0.02 0.004 0.006 2016 0.02 0.02 mg/L Endothal 0.1 0.045 0.094 2014 0.1 0.1 mg/L Endrin 0.002 0.0001 0.0003 2016 0.002 0.002 mg/L Ethylene dibromide (EDB) 5E-05 2E-05 1E-05 2003 5E-05 0 mg/L Glyphosate 0.7 0.025 0.9 2007 0.7 0.7 mg/L Heptachlor 1E-05 1E-05 8E-06 1999 0.0004 0 mg/L Heptachlor epoxide 1E-05 1E-05 6E-06 1999 0.0002 0 mg/L Hexachlorobenzene 0.001 0.0005 3E-05 2003 0.001 0 mg/L Hexachlorocyclopentadie

Just familiarize yourself with the dataset at this point. The only factors we will be considering for the remainder of this tutorial are the contaminants, state\_max\_level, and federal\_max\_level. The contaminant is pretty self-explanatory. The state maximum level is the maximum permissible amount allowed by the state of California, using the units in the last column. Similarly, the federal maximum level is that set by the US EPA. In this dataset, there are no null values in the state\_max\_level; however, this is not the case with the federal restrictions. 

### **Water Regulation Queries**

The first question we want to address:_  
How many_ federal standards are stricter (lower) than the California standards?

In [2]:
SELECT * 
FROM dbo.state_regulations
WHERE federal_max_level < state_max_level;

(0 rows affected)

Total execution time: 00:00:00.008

contaminant state_max_level state_detection_limit state_health_goal state_health_date federal_max_level federal_max_level_goal units

If you get an empty set, you are correct. There are no cases in which California has laxer standards than the federal government. 

Okay, next question:   
_How many_ state standards are stricter than the federal standards? 

This is slightly more complicated than it seems… If we consider only the contaminants where there is a federal standard, it’s a very easy script:

In [ ]:
SELECT  COUNT(contaminant) AS Number_of_Stricter_State_Maximums
FROM dbo.state_regulations
WHERE federal_max_level > state_max_level;

But what about the nulls? If there is a null federal standard, does that mean that the state is stricter in its regulations? 

If you think the answer to that question is yes, then we need to take an additional step to address this issue. We can use COALESCE to fill in the null federal values with the state value and then add a number so that the federal is automatically higher (less strict). 

Quick refresher on COALESCE — it will fill nulls in the target column with a reference column:

In [3]:
SELECT  contaminant, 
        state_max_level,
        federal_max_level,
        COALESCE(federal_max_level, state_max_level + 1) AS federal_not_null
FROM    dbo.state_regulations

(91 rows affected)

Total execution time: 00:00:00.017

contaminant state_max_level federal_max_level federal_not_null Aluminum 1 NULL 2 Antimony 0.006 0.006 0.006 Arsenic 0.01 0.01 0.01 Asbestos 7 7 7 Barium 1 2 2 Beryllium 0.004 0.004 0.004 Cadmium 0.005 0.005 0.005 Chromium, Total 0.05 0.1 0.1 Cyanide 0.15 0.2 0.2 Fluoride 2 4 4 Mercury 0.002 0.002 0.002 Nickel 0.1 NULL 1.1 Nitrate 10 10 10 Nitrite 1 1 1 Nitrate + Nitrite 10 NULL 11 Perchlorate 0.006 NULL 1.006 Selenium 0.05 0.05 0.05 Thallium 0.002 0.002 0.002 Copper 1.3 1.3 1.3 Lead 0.015 0.015 0.015 Gross Alpha Particle 15 15 15 Gross Beta Particle 4 4 4 Radium-226 + Radium-228 5 5 5 Strontium-90 8 NULL 9 Tritium 20000 NULL 20001 Uranium 20 30 30 Benzene 0.001 0.005 0.005 Carbon tetrachloride 0.0005 0.005 0.005 1,2-Dichlorobenzene 0.6 0.6 0.6 1,4-Dichlorobenzene(p-DCB) 0.005 0.075 0.075 1,1-Dichloroethane (1,1-DCA) 0.005 NULL 1.005 1,2-Dichloroethane (1,2-DCA) 0.0005 0.005 0.005 1,1-Dichloroethylene (1,1-DCE) 0.006 0.007 0.007 cis-1,2-Dichloroethylene 0.006 0.07 0.07 trans-1,2-Dichloroethylene 0.01 0.1 0.1 Dichloromethane (Methylene chloride) 0.005 0.005 0.005 1,2-Dichloropropane 0.005 0.005 0.005 1,3-Dichloropropene 0.0005 NULL 1.0005 Ethylbenzene 0.3 0.7 0.7 Methyl tertiary butyl ether (MTBE) 0.013 NULL 1.013 Monochlorobenzene 0.07 0.1 0.1 Styrene 0.1 0.1 0.1 1,1,2,2-Tetrachloroethane 0.001 0.1 0.1 Tetrachloroethylene (PCE) 0.005 0.005 0.005 Toluene 0.15 1 1 1,2,4-Trichlorobenzene 0.005 0.07 0.07 1,1,1-Trichloroethane (1,1,1-TCA) 0.2 0.2 0.2 1,1,2-Trichloroethane (1,1,2-TCA) 0.005 0.005 0.005 Trichloroethylene (TCE) 0.005 0.005 0.005 Trichlorofluoromethane (Freon 11) 0.15 NULL 1.15 1,1,2-Trichloro-1,2,2-Trifluoroethane (Freon 113) 1.2 NULL 2.2 Vinyl chloride 0.0005 0.002 0.002 Xylenes 1.75 10 10 Alachlor 0.002 0.002 0.002 Atrazine 0.001 0.003 0.003 Bentazon 0.018 NULL 1.018 Benzo(a)pyrene 0.0002 0.0002 0.0002 Carbofuran 0.018 0.04 0.04 Chlordane 0.0001 0.002 0.002 Dalapon 0.2 0.2 0.2 1,2-Dibromo-3-chloropropane (DBCP) 0.0002 0.0002 0.0002 2,4-Dichlorophenoxyacetic acid (2,4-D) 0.07 0.07 0.07 Di(2-ethylhexyl)adipate 0.4 0.4 0.4 Di(2-ethylhexyl)phthalate (DEHP) 0.004 0.006 0.006 Dinoseb 0.007 0.007 0.007 Diquat 0.02 0.02 0.02 Endothal 0.1 0.1 0.1 Endrin 0.002 0.002 0.002 Ethylene dibromide (EDB) 5E-05 5E-05 5E-05 Glyphosate 0.7 0.7 0.7 Heptachlor 1E-05 0.0004 0.0004 Heptachlor epoxide 1E-05 0.0002 0.0002 Hexachlorobenzene 0.001 0.001 0.001 Hexachlorocyclopentadiene 0.05 0.05 0.05 Lindane 0.0002 0.0002 0.0002 Methoxychlor 0.03 0.04 0.04 Molinate 0.02 NULL 1.02 Oxamyl 0.05 0.2 0.2 Pentachlorophenol 0.001 0.001 0.001 Picloram 0.5 0.5 0.5 Polychlorinated biphenyls (PCBs) 0.0005 0.0005 0.0005 Simazine 0.004 0.004 0.004 Thiobencarb 0.07 NULL 1.07 Toxaphene 0.003 0.003 0.003 1,2,3-Trichloropropane 5E-06 NULL 1.000005 2,3,7,8-TCDD (dioxin) 3E-08 3E-08 3E-08 2,4,5-TP (Silvex) 0.05 0.05 0.05 Total Trihalomethanes 0.08 0.08 0.08 Haloacetic Acids (five) (HAA5) 0.06 0.06 0.06 Bromate 0.01 0.01 0.01 Chlorite 1 1 1

To do this, we will require a subquery to first coalesce the data and then count the data from the parent query.

In [4]:
SELECT COUNT(subquery.contaminant) AS Number_of_Stricter_State_Maximums
FROM (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) > state_max_level
) AS subquery

(1 row affected)

Total execution time: 00:00:00.003

Number_of_Stricter_State_Maximums 42

As a reminder, using the subquery in the FROM clause requires an alias. Here I used subquery as a reminder, and I also specifically counted the contaminant, as some SQL varieties will allow subquery.\* but others do not. Since the purpose of this tutorial is to improve the clarity of code, the following is how I actually denote my subquery, so that when returning to the code, I know exactly what it was intended to do:

In [5]:
SELECT COUNT(Stricter_State_Maximums.contaminant) AS Number_of_Stricter_State_Maximums
FROM (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) > state_max_level
) AS Stricter_State_Maximums

(1 row affected)

Total execution time: 00:00:00.002

Number_of_Stricter_State_Maximums 42

### Using a Common Table Expression for Clarity

The WITH clause is used prior to the SELECT clause. The alias for the CTE will then be called in the FROM clause in the parent function. Compare this to the first subquery expression:

In [6]:
WITH cte AS (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) > state_max_level
)
SELECT COUNT(contaminant) AS Number_of_Stricter_State_Maximums
FROM cte

(1 row affected)

Total execution time: 00:00:00.003

Number_of_Stricter_State_Maximums 42

While I have seen many programmers just label these as cte1, cte2, etc., clarity matters! The same expression with better labeling:

In [ ]:
WITH Stricter_State_Maximums AS (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) > state_max_level
)
SELECT COUNT(contaminant) AS Number_of_Stricter_State_Maximums
FROM Stricter_State_Maximums

I promise, I did not intentionally set up the data such that 42 is the answer, but it should be, right? 

That was a simple example, where the use of a CTE may be a little clearer, but it’s not so different that it requires an entire new way of doing things… So let’s make this more complicated!

### **_Show the Number of Contaminants that have stricter state standards, stricter federal standards, and identical standards for state and federal:_**

Since we’re still in the WITH mind-frame, let’s start with that organization.

In [7]:
WITH 
Stricter_State_Maximums AS (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) > state_max_level
), 
Stricter_Federal_Maximums AS (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) < state_max_level
), 
Same_State_Federal_Maximums AS (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) = state_max_level
)
SELECT COUNT(contaminant) AS Number_of_Levels, 'State' AS Stricter_Restriction_Levels
FROM Stricter_State_Maximums
UNION ALL 
SELECT COUNT(contaminant), 'Federal' 
FROM Stricter_Federal_Maximums
UNION ALL 
SELECT COUNT(contaminant), 'Equal'
FROM Same_State_Federal_Maximums;

(3 rows affected)

Total execution time: 00:00:00.007

Number_of_Levels Stricter_Restriction_Levels 42 State 0 Federal 49 Equal

To write this, I used the previous query, only modifying the logical operator ( \< , \> , = ). I used UNION ALL to append the results to the same column and created a new column for the descriptive details. Alternatively, this query can be written to keep all results in a single row:

In [8]:
WITH 
Stricter_State_Maximums AS (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) > state_max_level
), 
Stricter_Federal_Maximums AS (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) < state_max_level
), 
Same_State_Federal_Maximums AS (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) = state_max_level
)
SELECT  COUNT(SSM.contaminant) AS Number_of_Stricter_State_Standards, 
        COUNT(SFM.contaminant) AS Number_of_Stricter_Federal_Standards, 
        COUNT(SM.contaminant) AS Number_of_Equal_Standards
FROM    dbo.state_regulations AS SR
LEFT JOIN Stricter_State_Maximums AS SSM
    ON SR.contaminant = SSM.contaminant
LEFT JOIN Stricter_Federal_Maximums AS SFM
    ON SR.contaminant = SFM.contaminant
LEFT JOIN Same_State_Federal_Maximums AS SM
    ON SR.contaminant = SM.contaminant

Warning: Null value is eliminated by an aggregate or other SET operation.

(1 row affected)

Total execution time: 00:00:00.027

Number_of_Stricter_State_Standards Number_of_Stricter_Federal_Standards Number_of_Equal_Standards 42 0 49

In this case, it did require joins, which can often be computationally costly. The following two queries are the corresponding subquery versions.

In [9]:
SELECT COUNT(contaminant) AS Count, 'State' AS Stricter_Restriction_Levels
FROM (
    SELECT  contaminant, 
            state_max_level,  
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) > state_max_level
) AS Stricter_State_Maximums
UNION ALL 
SELECT COUNT(contaminant), 'Federal'
FROM (
    SELECT  contaminant, 
            state_max_level, 
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) < state_max_level
) AS Stricter_Federal_Maximums
UNION ALL 
SELECT COUNT(contaminant), 'Equal'
FROM (
    SELECT  contaminant, 
            state_max_level,  
            COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
    FROM    dbo.state_regulations
    WHERE   COALESCE(federal_max_level, state_max_level + 1) = state_max_level
) AS Same_State_Federal_Maximums;

(3 rows affected)

Total execution time: 00:00:00.004

Count Stricter_Restriction_Levels 42 State 0 Federal 49 Equal

In [10]:
SELECT  COUNT(Stricter_State_Maximums.contaminant) AS Number_of_Stricter_State_Maximums,
        COUNT(Stricter_Federal_Maximums.contaminant) AS Number_of_Stricter_Federal_Maximums,
        COUNT(Same_State_Federal_Maximums.contaminant) AS Number_of_Same_State_Federal_Maximums
FROM dbo.state_regulations AS SR
    LEFT JOIN
    (
        SELECT  contaminant, 
                state_max_level, 
                COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
        FROM    dbo.state_regulations
        WHERE   COALESCE(federal_max_level, state_max_level + 1) > state_max_level
    ) AS Stricter_State_Maximums
    ON SR.contaminant = Stricter_State_Maximums.contaminant
    LEFT JOIN
    (
        SELECT  contaminant, 
                state_max_level, 
                COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
        FROM    dbo.state_regulations
        WHERE   COALESCE(federal_max_level, state_max_level + 1) = state_max_level
    ) AS Same_State_Federal_Maximums
        ON SR.contaminant = Same_State_Federal_Maximums.contaminant
    LEFT JOIN
    (
        SELECT  contaminant, 
                state_max_level, 
                COALESCE(federal_max_level, state_max_level + 1) AS federal_max_coalesced
        FROM    dbo.state_regulations
        WHERE   COALESCE(federal_max_level, state_max_level + 1) < state_max_level
    ) AS Stricter_Federal_Maximums
        ON SR.contaminant = Stricter_State_Maximums.contaminant

Warning: Null value is eliminated by an aggregate or other SET operation.

(1 row affected)

Total execution time: 00:00:00.007

Number_of_Stricter_State_Maximums Number_of_Stricter_Federal_Maximums Number_of_Same_State_Federal_Maximums 42 0 49

### **Nested Subqueries or WITH?**

The three remaining questions I would like to ask are as follows: 

1. **What percentage of federal restrictions are less strict (higher) than state restrictions?**
2. **What percentage of state restrictions are stricter (lower) than set federal restrictions?**
3. **What percentage of all contaminants have a stricter state maximum than federal, assuming a NULL value means it is a known contaminant that is unregulated at a federal level?**

While these three questions are seemingly asking the same question, each of them draws from a different subset of the provided information. The first question requires that all data drawn from have a federal standard. The second question is specifically asking for the values where there is a set federal standard as well. Only the third includes the standards set by the state where the federal standard is considered less strict, as it doesn’t exist. 

Let’s start with the subquery this time. The question I pose now: Where does this query actually start?

In [ ]:
SELECT  CAST(100*Laxer_Federal_Maximum_Count/ (SELECT COUNT(*) FROM dbo.state_regulations WHERE federal_max_level IS NOT NULL) AS DECIMAL(5,1)) AS Question_1,
        CAST(100*Stricter_State_Maximum_Count/ (SELECT COUNT(*) FROM (
            SELECT  contaminant, 
                    state_max_level, 
                    federal_max_level 
            FROM dbo.state_regulations 
            ) AS SSM) AS DECIMAL(5,1)) AS Question_2,
        CAST(100*Stricter_State_Coalesced_Count/ (SELECT COUNT(*) FROM (
            SELECT  contaminant, 
                    state_max_level, 
                    COALESCE(federal_max_level, state_max_level + 1) AS federal_not_null 
            FROM dbo.state_regulations) AS UFC 
            WHERE federal_not_null > state_max_level) AS DECIMAL(5,1)) AS Question_3
FROM (
    SELECT  COUNT(LFM.contaminant) AS Laxer_Federal_Maximum_Count, 
            COUNT(FM.contaminant) AS Federal_Maximum_Count, 
            COUNT(SSM.contaminant) AS Stricter_State_Maximum_Count, 
            COUNT(SSC.contaminant) AS Stricter_State_Coalesced_Count,
            COUNT(UFC.contaminant) AS Unregulated_Federal_Coalesced_Count
    FROM (
        SELECT  contaminant, 
                state_max_level, 
                federal_max_level 
        FROM dbo.state_regulations 
        WHERE federal_max_level IS NOT NULL) AS FM
        LEFT JOIN (
            SELECT  contaminant, 
                    state_max_level, 
                    federal_max_level 
            FROM dbo.state_regulations 
            WHERE federal_max_level > state_max_level) AS SSM 
        ON FM.contaminant = SSM.contaminant
        LEFT JOIN (
            SELECT  contaminant, 
                    state_max_level, 
                    COALESCE(federal_max_level, state_max_level + 1) AS federal_not_null 
            FROM dbo.state_regulations) AS UFC 
        ON FM.contaminant = UFC.contaminant
        LEFT JOIN (
            SELECT  contaminant, 
                    state_max_level, 
                    federal_max_level 
            FROM dbo.state_regulations 
            WHERE federal_max_level > state_max_level) AS SSC 
        ON UFC.contaminant = SSC.contaminant
        LEFT JOIN (
            SELECT  contaminant, 
                    state_max_level, 
                    federal_max_level 
            FROM (
                SELECT  contaminant, 
                        state_max_level, 
                        federal_max_level 
                FROM dbo.state_regulations 
                WHERE federal_max_level IS NOT NULL) AS FM 
            WHERE federal_max_level > state_max_level) AS LFM 
        ON UFC.contaminant = LFM.contaminant
) AS All_Counts;

**Hopefully using the WITH clause can help bring some clarity!**

In [11]:
WITH 
Unregulated_Federal_Coalesced AS (
    SELECT  contaminant, 
            state_max_level,
            COALESCE(federal_max_level, state_max_level + 1) AS federal_not_null
    FROM    dbo.state_regulations
), 
Stricter_State_Coalesced AS (
    SELECT  contaminant, 
            state_max_level,
            federal_not_null
    FROM    Unregulated_Federal_Coalesced
    WHERE   federal_not_null > state_max_level
),
Stricter_State_Maximums AS (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level
    FROM dbo.state_regulations
    WHERE federal_max_level > state_max_level
),
Federal_Maximums AS (
    SELECT  contaminant,
            state_max_level, 
            federal_max_level    
    FROM    dbo.state_regulations
    WHERE   federal_max_level IS NOT NULL
),
Laxer_Federal_Maximums AS (
    SELECT  contaminant, 
            state_max_level, 
            federal_max_level
    FROM    Federal_Maximums
    WHERE   federal_max_level > state_max_level
), 
All_Counts AS (
    SELECT  COUNT(LFM.contaminant) AS Laxer_Federal_Maximum_Count, 
            COUNT(FM.contaminant) AS Federal_Maximum_Count, 
            COUNT(SSM.contaminant) AS Stricter_State_Maximum_Count, 
            COUNT(SSC.contaminant) AS Stricter_State_Coalesced_Count,
            COUNT(UFC.contaminant) AS Unregulated_Federal_Coalesced_Count
    FROM    Unregulated_Federal_Coalesced AS UFC
        LEFT JOIN
            Stricter_State_Coalesced AS SSC
            ON UFC.contaminant = SSC.contaminant
        LEFT JOIN
            Stricter_State_Maximums AS SSM
            ON UFC.contaminant = SSM.contaminant
        LEFT JOIN 
            Federal_Maximums AS FM
            ON UFC.contaminant = FM.contaminant
        LEFT JOIN 
            Laxer_Federal_Maximums AS LFM
            ON UFC.contaminant = LFM.contaminant
)
SELECT  CAST(100*Laxer_Federal_Maximum_Count/ (SELECT Federal_Maximum_Count FROM All_Counts) AS DECIMAL(5,1)) AS Question_1, 
        CAST(100*Stricter_State_Maximum_Count/ (SELECT Unregulated_Federal_Coalesced_Count FROM All_Counts) AS DECIMAL(5,1)) AS Question_2,
        CAST(100*Stricter_State_Coalesced_Count/ (SELECT Unregulated_Federal_Coalesced_Count FROM All_Counts) AS DECIMAL(5,1)) AS Question_3
FROM    All_Counts

Warning: Null value is eliminated by an aggregate or other SET operation.

(1 row affected)

Total execution time: 00:00:00.027

Question_1 Question_2 Question_3 35.0 29.0 46.0

Starting at WITH, we then look at the unregulated\_federal\_coalesced expression that returns the contaminants, state maxes, and coalesced federal maxes, pulling those data from the state\_regulations table. The stricter\_state\_coalesced is pulling from the unregulated\_federal\_coalesced query, but further filters the data using the WHERE clause. This here is an example of a nested subquery that is organized linearly in an easy-to-comprehend manner. 

The final common table expression before the parent SELECT statement is the All\_Counts query, which joins each of the common table expression outputs to calculate the values. This reduces the need for nested subqueries in the parent function, which still requires\* subqueries to perform the percentage calculations. 

We final get our percentages for questions 1, 2, and 3:  35%, 29% and 46%

**\*\*\*** 

Does this final query really _require_ a subquery? No.

And hopefully you’re thinking, ‘surely there must be a better way to perform calculations without subqueries’. This is indeed the case, but as this tutorial is only introducing the concept of the WITH clause / common table expressions, I did not want to introduce additional advanced SQL concepts. 

### **\*\*\*** 

### **In Conclusion**

I hope it’s clear that from a readability standpoint, the code using Common Table Expressions (WITH) is much easier to follow and comprehend.

I hope that this was informative and easy to digest! My goal is to help others who have completed SQL coursework start to progress into more advanced SQL tactics to write more organized and better performing queries.